kotlinpoet/src/commonMain/kotlin/com/squareup/kotlinpoet/OriginatingElementsHolder.kt

```diff
@@ -15,24 +15,26 @@
  */
 package com.squareup.kotlinpoet
 
-import javax.lang.model.element.Element
+import com.squareup.kotlinpoet.jvm.alias.JvmDefaultWithCompatibility
+import com.squareup.kotlinpoet.jvm.alias.JvmElement
+import kotlin.jvm.JvmInline
 
 /** A type that can have originating [elements][Element]. */
 public interface OriginatingElementsHolder {
```

Comment:
It is implemented by many other types, e.g. `FunSpec`, `TypeSpec`. I think even though it's JVM-only, it still needs to be in the common

Reply:
The expect/actual mechanism allows actuals on the JVM to implement interfaces not exposed by the expect.

Here's an example in Okio:
- Expect: https://github.com/square/okio/blob/b6b4d52e5f0099328a5bc288249058e573dc19ff/okio/src/commonMain/kotlin/okio/Sink.kt#L45
- Actual: https://github.com/square/okio/blob/b6b4d52e5f0099328a5bc288249058e573dc19ff/okio/src/jvmMain/kotlin/okio/Sink.kt#L22

We should not expose JVM-only APIs in common. The need to create a ton of dummy typealiases to JVM types is a good indicator for an API that should not be in common.

Reply:
In the case of `FunSpec`, for example, if you want to use the expect/actual mechanism to make `FunSpec` in the JVM implement `OriginatingElementsHolder` in a way that other platforms don't, then `FunSpec` itself must be expect/actual (unless the extra in the interface is not abstract).

This can create a lot of duplicate logic or generate a lot of extra work. That's what I use typealias for: to reduce the amount of work due to compatibility.

If we don't accept any JVM-only typealias exposed in common, then almost all major types will become expect/actual, even though their logic doesn't really differ.

This is the same problem you mentioned in https://github.com/square/kotlinpoet/issues/304#issuecomment-1582560046, and I'm trying to get around it with typealias.

Exposing JVM-only types in common I think is a compromise for compatibility and maybe acceptable? 🤔 If it's unacceptable, that's fine too, except that later migrations to those major types will become more difficult.

> There doesn't seem to be any best of both worlds until [KT-20427](https://youtrack.jetbrains.com/issue/KT-20427/Allow-expect-declarations-with-a-default-implementation) is resolved?

Reply:
Would that require expect/actuals of `FunSpec` and co?

Reply:
If an interface is expect/actual and there is an extra abstract method in the JVM, then `FunSpec` must also be expect/actual. Otherwise, in common it can't know about that extra part of the JVM and thus can't implement it.